src/ir/auth_logic/ast_construction.cc

```diff
@@ -0,0 +1,164 @@
+//-------------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//-------------------------------------------------------------------------------
+
+#include <string>
+#include <vector>
+#include <utility>
+#include <stdexcept>
+#include <iostream>
+#include <memory>
+#include <variant>
+#include <optional>
+
+
+#include "antlr4-runtime.h"
+#include "absl/status/statusor.h"
+#include "src/common/utils/map_iter.h"
+#include "src/ir/auth_logic/ast.h"
+#include "src/ir/auth_logic/ast_construction.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicLexer.h" 
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicParser.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicVisitor.h"
+#include "src/ir/datalog/program.h"
+
+namespace raksha::ir::auth_logic {
+
+Principal Parser::ConstructPrincipal (AuthLogicParser::PrincipalContext* context) {
+    std::string name = context->ID()->getText();
+    return Principal(name);
```

Comment:
This final use of name should also use `std::move`, as `Principal` takes its `name` parameter by value and `name` is not used after this point in the function.

Alternatively, you could just inline the initializer of `name` into this reference like so:

```C++
return Principal(context->ID()->getText());
```